File: app/src/main/java/jp/juggler/subwaytooter/util/PostAttachment.java
package jp.juggler.subwaytooter.util;

import jp.juggler.subwaytooter.api.entity.TootAttachment;

public class PostAttachment {
	
	public interface Callback{
		void onPostAttachmentComplete(PostAttachment pa);
	}
	
	public static final int STATUS_UPLOADING = 1;
	public static final int STATUS_UPLOADED = 2;
	public static final int STATUS_UPLOAD_FAILED = 3;

	public int status;
	public TootAttachment attachment;
	public Callback callback;

	
	public PostAttachment(Callback callback){
		this.status = STATUS_UPLOADING;
		this.callback = callback;
	}
	public PostAttachment( TootAttachment a ){
		this.status = STATUS_UPLOADED;
		this.attachment = a;
	}
}

File: app/src/main/java/jp/juggler/subwaytooter/util/PostAttachment.kt
package jp.juggler.subwaytooter.util

import jp.juggler.subwaytooter.api.entity.TootAttachment
import kotlinx.coroutines.Job

class PostAttachment : Comparable<PostAttachment> {
    interface Callback {
        fun onPostAttachmentComplete(pa: PostAttachment)
        fun onPostAttachmentProgress()
    }

    enum class Status(val id: Int) {
        Progress(1),
        Ok(2),
        Error(3),
    }

    var isCancelled = false
    val job = Job()
    var status: Status
    var attachment: TootAttachment? = null
    var callback: Callback? = null
    var progress =""
        set(value){
            if( field!=value){
                field = value
                callback?.onPostAttachmentProgress()
            }
        }

    constructor(callback: Callback) {
        this.status = Status.Progress
        this.callback = callback
    }

    constructor(a: TootAttachment) {
        this.status = Status.Ok
        this.attachment = a
    }

    override fun compareTo(other: PostAttachment): Int {
        val ta = this.attachment
        val tb = other.attachment
        return if (ta != null) {
            if (tb == null) 1 else ta.id.compareTo(tb.id)
        } else {
            if (tb == null) 0 else -1
        }
    }
}
